design/dhcpless_ipam.md

```diff
@@ -0,0 +1,472 @@
+
+---
+apiVersion: metal3.io/v1alpha1
+kind: BareMetalHost
+metadata:
+  annotations:
+    baremetalhost.metal3.io/preprovisioningnetworkdata-reference: metal3preprovisioningdatatemplates.infrastructure.cluster.x-k8s.io/preprov-data-template
+...
+```
+
+In this flow there is no `MachineDeployment`, BareMetalHost resources are annotated to reference a specfic
+`Metal3PreprovisioningDataTemplate` which contains `networkData` which will be processed in the same way as
+described above, with the new controller generating a Secret from the template, and updaing the BareMetalHost
+`preprovisioningNetworkDataName`, at which point the BMO starts inspection and other pre-provisioning operations.
+
+The preprovisioningNetworkDataName is used by default for networkData in the baremetal-operator, so in this configuration it's not
+strictly necessary to specify networkData via Metal3DataTemplate, however we'll want to delete the IPClaim after preprovisioning
+in the decoupled flow below so it seems likely we'll want to behave consistently and rely on the IP Reuse functionality if a
+consistent IP is required between pre-provisioning and provisioning phases.
+
+#### Preprovisioning Decoupled IPPool
+
+```yaml
+apiVersion: ipam.metal3.io/v1alpha1
+kind: IPPool
+metadata:
+  name: pool-1
+spec:
+  clusterName: cluster
+
+---
+
+apiVersion: ipam.metal3.io/v1alpha1
+kind: IPPool
+metadata:
+  name: preprovisioning-pool
+
+---
+
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: Metal3PreprovisioningDataTemplate
+metadata:
+  name: preprov-data-template
+spec:
+  preprovisioningNetworkData:
+    networks:
+      ipv4:
+      - id: eth0
+        ipAddressFromIPPool: preprovisioning-pool
+
+---
+
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: Metal3DataTemplate
+metadata:
+  name: data-template
+spec:
+  clusterName: cluster
+  networkData:
+    networks:
+      ipv4:
+      - id: eth0
+        ipAddressFromIPPool: pool-1
+
+---
+
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: Metal3MachineTemplate
+metadata:
+  name: machine-template
+spec:
+  template:
+    spec:
+      dataTemplate:
+        name: data-template
+      hostSelector:
+        matchLabels:
+          cluster-role: control-plane
+
+---
+apiVersion: metal3.io/v1alpha1
+kind: BareMetalHost
+metadata:
+  annotations:
+    baremetalhost.metal3.io/preprovisioningnetworkdata-reference: metal3preprovisioningdatatemplates.infrastructure.cluster.x-k8s.io/preprov-data-template
+...
+
+```
+
+In this flow we have `preprovisioning-pool` which is not associated with any cluster, this is used to provide an IPAddress during
+the pre-provisioning phase as described above.  To reduce the required size of the pool, the IPClaim will be deleted after the
+preprovisioning phase is completed, e.g the BMH resource becomes available.
+
+In the provisioning phase another pool, associated with a cluster is used to template networkData as in the existing process.
```

Comment:
The first use-case `Preprovisioning - Common IPPool` covers the case where the same network configuration is used for IPA and during provisioning, I agree this is the simpler and probably more common case.

My understanding is there are also a use-case where a separate network is used for the pre-provisioning phase, so I'm trying to cover that scenario here.

I agree perhaps reclaiming the IPClaim after inspection introduces undue complexity, but if we don't do that the provisioning network address range potentially becomes a lot larger (number of deployed BMH's vs the concurrency allowed by BMO), perhaps that's a worthwhile tradeoff and we can consider the IPClaim deletion in future if it proves necessary?

Reply:
The use case where there'd be two is if you have hosts that are dynamically assigned to clusters (and the networks reconfigured at the same time). I'm not sure this exists in the wild, but it would be good to allow it.
In the more common case of the network staying the same, I think you really want the IP address not to change. The one in the inspection data is the one you want to have in the cluster. So I think ideally you only want to delete the preprovisioning IP claim at provisioning time, and only if a separate provisioned IP claim exists.

Reply:
You can support dynamically assigning hosts to clusters without changing the IPA network configuration. It is true that the network used by Kubernetes needs to be reconfigured and be visible from the capm3 controller. On the other hand the IPA  only needs to see Ironic (and os image repository). Servers should not see each other. As the requirements are really different, the best solution is to keep two separate network configurations for IPA and CAPM3 and in that case ipclaims are on different pools. If you have cluster definitions and baremetalhosts on different clusters (as with #429 or #408) I am not sure there is an easy path to keep ipclaims synchronized. If you really need to do something like that with a single network (that is what we were doing in old versions of Kanod and yes it was a bad idea), I would put the ipam on the baremetalHost side and push  IPs through BaremetalHost annotations and that is why in our "template for BMH" experiment [BmhData](https://gitlab.com/Orange-OpenSource/kanod/bmh-data), metadata could also be exported as BMH annotations.

Reply:
I might be wrong but this thread seem to be started with a misunderstanding. The question was:
"you want to use two network configuration for the IPA?" the answer is **NO**.

Then the discussion was mixed with reclaiming after inspection and @zaneb has correctly pointed out that pre-provisioning lasts until the BMH is in "provisioned" state, by default this means the reboot of the physical machine.

IMO the pre-provisioning IP can be reclaimed/released after the BMH is in "provisioned" state.

To summarize:

IPA will use only 1 IPPool during its runtime (preprovisioning pool), the second IPPool is used to generate the network data of the provisioned node and  that network data is just written to the disk into the config-drive partition as network data that will be used by the provisioned machine after it restarts.

I would also like to avoud discussion about syncronizing the pools and multy tenancy as this proposal is simply about implementing pre-provisioning network data generation and linking it to the BMH. 

I might be saying things that you already know but for me at least it feels this thread has been a bit derailed from the proposal.